### PR TITLE
[vshare] extractor rewritten

### PR DESCRIPTION
**Pull request by timendum**
### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [X] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

The VShare extractor is broken, I've rewritten it to depack the code in the page and get the video url from there.
I've also added the `_extract_urls` method.

This should close issue #14473.

### CODE REVIEW

**Comment by timendum** [comment]
@dstftw Can I do something to improve this PR?

File: youtube_dl/extractor/vshare.py [hunk 1]
```diff
@@ -19,20 +23,38 @@ class VShareIE(InfoExtractor):
...
+        title = self._html_search_regex(r'<title>([^<]+)</title>', webpage, 'title')
+        title = title.split(' - ')[0]
 
-        title = self._html_search_regex(
-            r'(?s)<div id="root-container">(.+?)<br/>', webpage, 'title')
-        video_url = self._search_regex(
-            r'<a[^>]+href=(["\'])(?P<url>(?:https?:)?//.+?)\1[^>]*>[Cc]lick\s+here',
-            webpage, 'video url', group='url')
+        unpacked = self._extract_packed(webpage)
+        video_url = self._search_regex(r'<source src="([^"]+)', unpacked, 'video url')
```
Comment (dstftw) [inline youtube_dl/extractor/vshare.py]: All formats should be extracted.

File: youtube_dl/extractor/vshare.py [hunk 2]
```diff
@@ -19,20 +23,38 @@ class VShareIE(InfoExtractor):
...
 
         return {
             'id': video_id,
             'title': title,
             'url': video_url,
         }
+
+    @staticmethod
+    def _extract_urls(webpage):
+        print(webpage)
```
Comment (dstftw) [inline youtube_dl/extractor/vshare.py]: ^

File: youtube_dl/extractor/generic.py [hunk 3]
```diff
@@ -2879,6 +2880,11 @@ def _real_extract(self, url):
             return self.playlist_from_matches(
                 channel9_urls, video_id, video_title, ie=Channel9IE.ie_key())
 
+        vshare_urls = VShareIE._extract_urls(webpage)
```
Comment (dstftw) [inline youtube_dl/extractor/generic.py]: Add a test.

File: youtube_dl/extractor/vshare.py [hunk 4]
```diff
@@ -19,20 +23,38 @@ class VShareIE(InfoExtractor):
...
 
+    def _extract_packed(self, webpage):
+        packed = self._search_regex(r'(eval\(function.+)', webpage, 'packed code')
+        unpacked = decode_packed_codes(packed)
+        digits = self._search_regex(r'\[((?:\d+,?)+)\]', unpacked, 'digits')
+        digits = digits.split(',')
+        digits = [int(digit, 10) for digit in digits]
+        key_digit = self._search_regex(r'fromCharCode\(.+?(\d+)\)}', unpacked, 'key digit')
+        key_digit = int(key_digit, 10)
+        chars = [chr(d - key_digit) for d in digits]
```
Comment (dstftw) [inline youtube_dl/extractor/vshare.py]: `compat_chr`.

File: youtube_dl/extractor/vshare.py [hunk 5]
```diff
@@ -19,20 +23,38 @@ class VShareIE(InfoExtractor):
         'only_matching': True,
     }]
 
+    def _extract_packed(self, webpage):
+        packed = self._search_regex(r'(eval\(function.+)', webpage, 'packed code')
+        unpacked = decode_packed_codes(packed)
+        digits = self._search_regex(r'\[((?:\d+,?)+)\]', unpacked, 'digits')
+        digits = digits.split(',')
+        digits = [int(digit, 10) for digit in digits]
```
Comment (dstftw) [inline youtube_dl/extractor/vshare.py]: 10 is default.

**Comment by timendum** [comment]
@dstftw thanks for the feedback (and sorry for the `print`).

It should be ok now.

Timendum

**Pull request merged**